java/code/src/com/suse/manager/reactor/hardware/HardwareMapper.java

```diff
@@ -664,19 +671,36 @@ else if (smbiosRecordsSystem.isPresent()) {
                             null, server, vCPUs, memory);
                 }
                 else {
+                    boolean isForeign = virtualInstance.getHostSystem() != null &&
+                            virtualInstance.getHostSystem().hasEntitlement(EntitlementManager.FOREIGN);
+                    long newMemory = memory;
+                    // Only foreign system (s390 and VHM) and systems with no memory set should have updated memory
+                    if (!isForeign && virtualInstance.getTotalMemory() != 0 &&
```

Comment:
If `virtualInstance.getTotalMemory()` is `null`, then the comparison `virtualInstance.getTotalMemory() != 0` throws a NPEx. I think the order of those comparisons should be flipped.

Moreover: that the code in this class has grown into a monster over the time :smile: IMO extracting this complex check in a separate method (e.g. `boolean shouldUpdateForeignSystemMemory(VirtualInstance vi)`) could be slightly more clear.